### PR TITLE
Ant external adapter to launch error-prone in a separate process

### PR DESCRIPTION
**Pull request by garricko**
This adds an Ant compiler adapter that can work in forked mode. Should be the same as pull request 270 but without the messed up rebase.

### CODE REVIEW

File: ant/src/main/java/com/google/errorprone/ErrorProneExternalCompilerAdapter.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+import org.apache.tools.ant.types.Commandline.Argument;
+import org.apache.tools.ant.types.Path;
+import org.apache.tools.ant.types.Reference;
+import org.apache.tools.ant.util.JavaEnvUtils;
+import org.apache.tools.ant.util.LoaderUtils;
+
+/**
+ * Ant component to launch an external javac with error-prone enabled.
+ */
+public class ErrorProneExternalCompilerAdapter extends DefaultCompilerAdapter {
```
Comment (cushon) [inline ant/src/main/java/com/google/errorprone/ErrorProneExternalCompilerAdapter.java]: Does this work with JDK7?

File: ant/src/main/java/com/google/errorprone/ErrorProneExternalCompilerAdapter.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        classpath = new Path(getProject());
+      }
+      // Usually redundant, but check two resources in case Ant stuff is in a different jar
+      addResourceSource(classpath, "com/google/errorprone/ErrorProneExternalCompilerAdapter.class");
+      addResourceSource(classpath, "com/google/errorprone/ErrorProneCompiler.class");
+      addResourceSource(classpath, "com/sun/tools/javac/Main.class");
+      cmd.createArgument().setPath(classpath);
+      cmd.createArgument().setValue(ErrorProneCompiler.class.getName());
+      if (suggestFixes) {
+        cmd.createArgument().setValue("-Xjcov");
```
Comment (cushon) [inline ant/src/main/java/com/google/errorprone/ErrorProneExternalCompilerAdapter.java]: We don't need to pass `-Xjcov` any more.

File: examples/ant/build.xml [hunk 3]
```diff
@@ -15,10 +15,22 @@
...
+      <errorprone suggestfixes="yes"/>
+    </javac>
+  </target>
+  <target name="compile.internal">
+    <delete dir="build"/>
+    <mkdir dir="build"/>
+    <!--NOTE: error_prone_ant.jar must be installed in ANT_HOME/lib for the internal compiler -->
     <javac srcdir="src" destdir="build" includeantruntime="false"
            compiler="com.google.errorprone.ErrorProneAntCompilerAdapter">
+      <compilerarg value="-Xjcov"/> <!-- for suggested fixes, increases compiler memory requirements -->
```
Comment (cushon) [inline examples/ant/build.xml]: `-Xjcov` can be removed here also.

File: examples/ant/build.xml [hunk 4]
```diff
@@ -15,10 +15,22 @@
   -->
 <project name="error prone example" default="compile">
   <target name="compile">
+    <delete dir="build"/>
     <mkdir dir="build"/>
-    <!--NOTE: error_prone_ant.jar must be installed in ANT_HOME/lib -->
+    <!-- For external compiles (fork="yes") you can do it this way -->
+    <componentdef name="errorprone" classname="com.google.errorprone.ErrorProneExternalCompilerAdapter"
+                  classpath="../../ant/target/error_prone_ant-1.1.3-SNAPSHOT.jar"/>
```
Comment (cushon) [inline examples/ant/build.xml]: This path is stale.

File: ant/src/main/java/com/google/errorprone/ErrorProneExternalCompilerAdapter.java (hunk 1 above)
Comment (garricko) [inline ant/src/main/java/com/google/errorprone/ErrorProneExternalCompilerAdapter.java]: I have been using it with JDK7 without issue.

File: ant/src/main/java/com/google/errorprone/ErrorProneExternalCompilerAdapter.java (hunk 2 above)
Comment (garricko) [inline ant/src/main/java/com/google/errorprone/ErrorProneExternalCompilerAdapter.java]: Did that change recently (in those last 88 changes)? When I initially created this it did seem to change the behavior, though it seemed a little unreliable.

File: ant/src/main/java/com/google/errorprone/ErrorProneExternalCompilerAdapter.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
+package com.google.errorprone;
```
Comment (cushon) [inline ant/src/main/java/com/google/errorprone/ErrorProneExternalCompilerAdapter.java]: Please include the standard license header.

File: ant/src/main/java/com/google/errorprone/ErrorProneExternalCompilerAdapter.java (hunk 2 above)
Comment (cushon) [inline ant/src/main/java/com/google/errorprone/ErrorProneExternalCompilerAdapter.java]: Yep, it changed in 9265d0fe23deb0132161d09408e01b3269047960

**Comment by cushon** [comment]
Thanks!

**Pull request merged**